Supplement: Add and update data-test for new add page

jerolimov opened the pull request:
**Fixes**: 
Epic: https://issues.redhat.com/browse/ODC-5013

**Analysis / Root cause**:
Data-test attributes are missed for getting started card and items.

**Solution Description**: 
* Update add card `data-test-id`s to hierarchy-based `data-test` attributes.
* Add missing `data-test` attributes to new getting started section.
* Updated `packages/dev-console/integration-tests/support/pages/add-flow/add-page.ts` with new data-test checks

**Screen shots**: 
Some `data-test` query examples:

```js
// Remove outlines for all data-test and data-test-id entries
document.querySelectorAll('[data-test],[data-test-id]').forEach(e => e.style.outline = '')
// Add outline to all add-page data-test elements
document.querySelectorAll('[data-test=add-page] [data-test]').forEach(e => e.style.outline = '1px solid red')
```
![image](https://user-images.githubusercontent.com/139310/119571131-45a16c00-bdb1-11eb-8733-d1036bc35021.png)

```js
// Remove outlines for all data-test and data-test-id entries
document.querySelectorAll('[data-test],[data-test-id]').forEach(e => e.style.outline = '')
// Add outline to all add-page titles
document.querySelectorAll('[data-test=add-page] [data-test~=title]').forEach(e => e.style.outline = '1px solid red')
```
![image](https://user-images.githubusercontent.com/139310/119570928-0246fd80-bdb1-11eb-963f-c7ea2326673a.png)


**Unit test coverage report**: 
* Updated existing tests. Don't added any `data-test` tests.

**Test setup:**
* Open Add page and check rendered elements.

**Browser conformance**: 
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

christianvogt commented on frontend/packages/console-shared/src/components/getting-started/GettingStartedCard.tsx:
```diff
@@ -51,8 +53,14 @@ export const GettingStartedCard: React.FC<GettingStartedCardProps> = ({
       direction={{ default: 'column' }}
       grow={{ default: 'grow' }}
       className="ocs-getting-started-card"
+      data-test-id={`ocs-getting-started-card-${id}`}
```
Use space separated values so we can create selectors with `~=`

```suggestion
      data-test-id={`ocs-getting-started-card ${id}`}
```

christianvogt commented on frontend/packages/console-shared/src/components/getting-started/GettingStartedCard.tsx:
```diff
@@ -51,8 +53,14 @@ export const GettingStartedCard: React.FC<GettingStartedCardProps> = ({
+        headingLevel="h3"
+        size={TitleSizes.md}
+        style={{ color: titleColor }}
+        data-test-id="ocs-getting-started-card-title"
```
Here we don't really need fully qualified IDs because we should be first selecting the card before selecting any of its children.
We should discuss this as part of the conventions with the team.

My first thought is that this could simply be `data-test-id="title"`

christianvogt commented on frontend/packages/console-shared/src/components/getting-started/GettingStartedCard.tsx:
```diff
@@ -64,17 +72,25 @@ export const GettingStartedCard: React.FC<GettingStartedCardProps> = ({
+                          href: link.href,
+                          target: '_blank',
+                          rel: 'noopener noreferrer',
+                          'data-test-id': `ocs-getting-started-card-item-${link.id}`,
```
Use space separated values:
```suggestion
                          'data-test-id': `ocs-getting-started-card-item ${link.id}`,
```

nemesis09 commented:
/lgtm

nemesis09 commented:
/lgtm cancel

nemesis09 commented:
Sorry, the comments did not appear when I added lgtm, maybe loading issues at my end. cancelled lgtm.

christianvogt commented:
/hold for meeting to discuss conventions

jerolimov commented:
/assign @makambalaji @christianvogt 
/cc @makambalaji @gajanan-more @sanketpathak 

Hi all, I updated this PR based on our discussion today. Please take a look and let me know what you think.

Initial I started this PR to add `data-test` attributes to the new getting started section. I now also updated the `data-test-id`s on the "complete" add page, which means for the new add cards.

I switched also from `data-test-id` to `data-test` after I find this resources:
* [frontend/packages/integration-tests-cypress/README.md](https://github.com/openshift/console/blob/master/frontend/packages/integration-tests-cypress/README.md) (see Best Practices # 2)
* https://docs.cypress.io/guides/references/best-practices#Real-World-Example
* https://testing-library.com/docs/queries/bytestid/

If you want I can check tomorrow also if I could update the existing add page integration-tests. I don't run them now, but I expect that the tests fails after the redesign or at least with this data-test changes here, right?

jerolimov commented on frontend/packages/console-shared/src/components/getting-started/GettingStartedCard.tsx:
```diff
@@ -51,30 +53,43 @@ export const GettingStartedCard: React.FC<GettingStartedCardProps> = ({
       direction={{ default: 'column' }}
       grow={{ default: 'grow' }}
       className="ocs-getting-started-card"
+      data-test={`column ${id}`}
```
I wasn't sure how to call this three columns.

The component is called card (which was maybe a mistake), but I added a "card" data-test attribute to complete "getting started" card. This matches more the visual part of the component and not the component name.

christianvogt replied:
Perhaps this class should have been named `Item` or `Entry` instead of `Card`.

gajanan-more replied:
Can you please change it to `data-test-id={`column ${id}`}`

jerolimov replied:
Yep, renaming the `GettingStartedGrid` to `GettingStartedCard` and `GettingStartedCard` to `GettingStartedItem` make sense. But if we would rename this to `GettingStartedItem` this could confuse because the links are also called items internally and use `data-test="item"`

So fetching all items would return the columns and the links. A downside of unprefixed, general terms like "item" :confused: I add some more examples to verify here if our concept will work well on this "simple" page.

Currently:
* GettingStartedSection > 1 grid > n cards > n items

Options I see
1. GettingStartedSection > 1 card > n columns > n items (IMHO fine)
2. GettingStartedSection > 1 card > n items > n items (IMHO this will make problems)
3. GettingStartedSection > 1 card > n items > n actions (IMHO fine)
4. GettingStartedSection > 1 card > n items > n options (sounds like a dropdown, but we use this term in the UI)

We can align this also with the AddCards

Currently:
* Add card section > n cards > n items

Options I see:
1. Keep cards and items
2. cards and actions
3. cards and options

(Solved the `data-test` change request with Gajanan in slack.)

jerolimov replied:
It is `column` now.

jerolimov replied:
Changed the `data-test` attribute for now to `card ${id}` so that this matches also our code structure.

I also added a data-test wrapper for the getting started section and the add cards. So its possible to distinguish them via:

```
[data-test~=getting-started] [data-test~=card $id]
or for all getting started cards (up to 3 at the moment)
[data-test~=getting-started] [data-test~=card]
```

or

```
[data-test~=add-cards] [data-test~=card $id]
or for all add cards:
[data-test~=add-cards] [data-test~=card]
```

For the getting started section, this works for example via `cy.byTestID('card samples')`,  `cy.byTestID('card quick-start')` and  `cy.byTestID('card developer-features')` for the developer getting started sections.

/cc @gajanan-more

jerolimov commented on frontend/packages/console-shared/src/components/getting-started/GettingStartedCard.tsx:
```diff
@@ -94,16 +109,28 @@ export const GettingStartedCard: React.FC<GettingStartedCardProps> = ({
+              onClick={moreLink.onClick}
+              isInline
+              variant="link"
+              data-test={`more-item ${moreLink.id}`}
```
Is it a `more-item` or also just a `item`?

jerolimov replied:
Changed it to just `item`.

jerolimov commented on frontend/packages/dev-console/src/components/add/AddPageLayout.tsx:
```diff
@@ -54,10 +54,13 @@ const AddPageLayout: React.FC<AddPageLayoutProps> = ({ title, hintBlock: additio
                 }}
-                data-test-id="odc-add-page-details-switch"
+                data-test="details-switch"
               />
             </Tooltip>
-            <span className="odc-add-page-layout__hint-block__details-switch__text">
+            <span
+              className="odc-add-page-layout__hint-block__details-switch__text"
+              data-test="details-label"
+            >
```
`details-switch` and `details-label` doesn't match our meeting result.

Should we add a div here which has data-test="details" and then "switch" and "label" to the child elements?

jerolimov replied:
Changed it to

`details-options` > `switch`
`details-options` > `label`

christianvogt commented on frontend/packages/console-shared/src/components/getting-started/GettingStartedCard.tsx:
```diff
@@ -94,16 +109,28 @@ export const GettingStartedCard: React.FC<GettingStartedCardProps> = ({
+              onClick={moreLink.onClick}
+              isInline
+              variant="link"
+              data-test={`item ${moreLink.id}`}
```
My first thought was that this should be a separate id from the list items.

jerolimov replied:
Wouldn't you describe the UI that each column contains three links? The idea here is that a test can assert for 3 links (items).

christianvogt replied:
In this case I have no strong opinion. It comes down to how we want to test it.

christianvogt commented on frontend/packages/console-shared/src/components/getting-started/GettingStartedGrid.tsx:
```diff
@@ -55,10 +56,10 @@ export const GettingStartedGrid: React.FC<GettingStartedGridProps> = ({ onHide,
 
   return (
-    <Card className="ocs-getting-started-grid">
+    <Card className="ocs-getting-started-grid" data-test="card getting-started">
```
This ID could clash with the dynamic IDs of the add actions cards.

jerolimov replied:
I think that's a feature, not a bug. If someone creates a dynamic "Getting Started" card, it's okay that our **tests** fail.

"For the tests" (visual) this is a card.

jerolimov replied:
Removed the card for now. So tests need to check for `getting-started`. /cc @gajanan-more

christianvogt commented on frontend/packages/dev-console/src/components/add/AddCard.tsx:
```diff
@@ -4,18 +4,19 @@ import { ResolvedExtension, AddAction } from '@console/dynamic-plugin-sdk';
   const isTitleFromItem: boolean = items?.length === 1 && items[0].properties.label === title;
   return items?.length > 0 ? (
-    <Card key={title} className="odc-add-card">
+    <Card key={title} className="odc-add-card" data-test={`card ${id}`}>
```
This dynamic ID could clash the static `card getting-started` ID within the same scope.

jerolimov replied:
I think that's a feature, not a bug. If someone creates a dynamic "Getting Started" card, it's okay that our **tests** fail.

"For the tests" (visual) this is a card.

christianvogt replied:
But add-actions extensions and the getting start card are 2 separate contributed items. One is static and one is dynamic. And there is no scope difference. So while each card is visually distinguishable and valid, from a test perspective the test would need to change and not the implementation of the contributions.

jerolimov replied:
Removed the card from the `getting-started` (data-test) card.

jerolimov commented:
> /hold for meeting to discuss conventions

The current PR matches now the latest `data-test` discussion
/unhold

PTAL.

jerolimov commented:
> /hold for meeting to discuss conventions

The current PR matches now the latest `data-test` discussion
/hold cancel

PTAL.

gajanan-more reviewed (commented):
For now, I feel it will work for us and will make automation more robust. @makambalaji Do you have any inputs? IF not, we are good to go. Thanks @jerolimov

makambalaji commented on frontend/packages/dev-console/integration-tests/support/pages/add-flow/add-page.ts:
```diff
@@ -10,61 +10,61 @@ export const addPage = {
       case 'Git':
       case addOptions.Git:
-        cy.byLegacyTestID('odc-add-card-item-import-from-git').click();
+        cy.byTestID('item import-from-git').click();
```
Based on the current DOM structure, `data-test-id="odc-add-card-item-import-from-git"`. 

To use byTestID, data-* attribute needs to change as `data-test="odc-item import-from-git"` Have we updated this as part of any PR?

And also currently this attribute is present on div element. instead it should be on link to use "click" action

jerolimov replied:
Hey @makambalaji, I tested this by running (some) cypress tests incl. some which selects an "add card option" (or action, or item..) and they works fine. See attached screenshot:

![Screenshot from 2021-05-31 15-34-20](https://user-images.githubusercontent.com/139310/120216312-27b38b80-c237-11eb-972e-67c82ffaf5ce.png)

But, as we discussed earlier, I changed the structure of the add page card item to remove the `div` element between the `ul` and `li` element. /cc @nemesis09 Please see latest change https://github.com/openshift/console/compare/bee76ef4ab041e0ab0ac77a766c385f88567a130..2094faa4f6b0beca8b34ca2be6aa8c53c8b1f6f2

Before:
![Screenshot from 2021-05-31 17-32-57](https://user-images.githubusercontent.com/139310/120216903-e1126100-c237-11eb-80e8-56d88b7d949a.png)

After:
![Screenshot from 2021-05-31 17-32-26](https://user-images.githubusercontent.com/139310/120216897-deb00700-c237-11eb-9c91-e410276bebfb.png)

So, this does not not exactly address your issue. But because the Add Card works with an `onClick` element, I expect all works fine for the cypress tests with this change.

nemesis09 commented:
tested locally. works as expected.
/lgtm

jerolimov commented:
/test e2e-gcp-console
